hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java

```diff
@@ -68,139 +75,216 @@
+    protected void restoreFromSnapshot(@Nonnull Object key, @Nonnull Object value) {
+        int partition = ((BroadcastKey<Integer>) key).key();
+        long offset = (Long) value;
+
+        if (assignedPartitions.contains(partition)) {
+            // after snapshot restore, we try to continue where we left off. However the
+            // current oldest sequence might be greater than the restored offset. This would cause
+            // stale sequence exception, so we fast forward to the oldest sequence instead.
+            long oldest = subscriptions.get(partition).getOldestSequence();
+            long newOffset = Math.max(oldest, offset);
+            if (newOffset != offset) {
+                getLogger().warning("Events lost for partition " + partition + " when restoring from " +
+                        "snapshot. Requested was: " + offset + ", current head is: " + oldest);
+            }
+            readOffsets.put(partition, newOffset);
+            emittedOffsets.put(partition, newOffset);
         }
+    }
+
+    @Override
+    public boolean finishSnapshotRestore() {
+        logFinest(getLogger(), "Restored snapshot. Offsets: %s", readOffsets);
+        return true;
+    }
+
+    @Override
+    public boolean isCooperative() {
         return false;
     }
 
-    private void nextTraverserIfNeeded() {
-        if (traverser != null) {
-            return;
+    private void initialRead() {
+        // readOffsets and emittedOffsets are not empty if they were restored from snapshot
+        if (readOffsets.isEmpty()) {
+            subscriptions.forEach((partition, future) ->
+                    uncheckRun(() -> readOffsets.put(partition, getSequence(subscriptions.get(partition))))
+            );
+            emittedOffsets.putAll(readOffsets);
+        }
+        readFutures = readOffsets.entrySet().stream()
+            .collect(toMap(Entry::getKey, e -> readFromJournal(e.getKey(), e.getValue())));
+    }
+
+    private long getSequence(EventJournalInitialSubscriberState state) {
+        return startFromNewest ? state.getNewestSequence() + 1 : state.getOldestSequence();
+    }
+
+    private Traverser<T> nextTraverser() {
+        ReadResultSet<T> resultSet = nextResultSet();
+        if (resultSet == null) {
+            return null;
         }
-        ReadResultSet<T> resultSet = null;
-        Iterator<Entry<Integer, ICompletableFuture<ReadResultSet<T>>>> iterator = futureMap.entrySet().iterator();
-        while (iterator.hasNext() && resultSet == null) {
+        Traverser<T> traverser = traverseIterable(resultSet);
+        final ReadResultSet<T> currentSet = resultSet;
+        return peekIndex(traverser, i -> {
+            pendingOffset = currentSet.getSequence(i) + 1;
+        });
+    }
+
+    private ReadResultSet<T> nextResultSet() {
+        Iterator<Entry<Integer, ICompletableFuture<ReadResultSet<T>>>> iterator = readFutures.entrySet().iterator();
+        while (iterator.hasNext()) {
```

Comment:
This is prone to starvation. If the first future always manages to be done since the last call, for example when the downstream is slow, we won't touch another futures.